runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+   * and Timers API</a> once it is available.
+   */
+  private static class ProcessFn<
+          InputT, RestrictionT, OutputT, TrackerT extends RestrictionTracker<RestrictionT>>
+      extends OldDoFn<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+      implements OldDoFn.RequiresWindowAccess {
+    /**
+     * The state cell containing a watermark hold for the output of this {@link DoFn}. The hold is
+     * acquired during the first {@link DoFn.ProcessElement} call for this element and restriction,
+     * and is released when the {@link DoFn.ProcessElement} call returns {@link
+     * DoFn.ProcessContinuation#stop}.
+     *
+     * <p>A hold is needed to avoid letting the output watermark immediately progress together with
+     * the input watermark when the first {@link DoFn.ProcessElement} call completes.
+     *
+     * <p>The hold is updated with the future output watermark reported by ProcessContinuation.
+     */
+    private static final StateTag<Object, WatermarkHoldState<GlobalWindow>> watermarkHoldTag =
+        StateTags.makeSystemTagInternal(
+            StateTags.<GlobalWindow>watermarkStateInternal(
+                "hold", OutputTimeFns.outputAtLatestInputTimestamp()));
+
+    /**
+     * The state cell containing a copy of the element. Written during the first {@link
+     * DoFn.ProcessElement} call and read during subsequent calls in response to timer firings, when
+     * the original element is no longer available.
+     */
+    private final StateTag<Object, ValueState<WindowedValue<InputT>>> elementTag;
+
+    /**
+     * The state cell containing a restriction representing the unprocessed part of work for this
+     * element.
+     */
+    private StateTag<Object, ValueState<RestrictionT>> restrictionTag;
+
+    private final DoFn<InputT, OutputT> fn;
+    private transient DoFnInvoker<InputT, OutputT> invoker;
+
+    ProcessFn(
+        DoFn<InputT, OutputT> fn,
+        Coder<InputT> elementCoder,
+        Coder<? extends BoundedWindow> windowCoder) {
+      this.fn = fn;
+      elementTag =
+          StateTags.value("element", WindowedValue.getFullCoder(elementCoder, windowCoder));
+    }
+
+    @Override
+    public void setup() throws Exception {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+      restrictionTag =
+          StateTags.value("restriction", invoker.<RestrictionT>invokeGetRestrictionCoder());
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
+      final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT> internals =
+          c.windowingInternals();
+      StateInternals<?> stateInternals = internals.stateInternals();
+      TimerInternals timerInternals = internals.timerInternals();
+
+      ValueState<WindowedValue<InputT>> elementState =
+          stateInternals.state(StateNamespaces.global(), elementTag);
```

Comment:
I could use c.window() after https://github.com/apache/incubator-beam/pull/863 is merged. Currently I get an exception from c.window() if this is a processElement() call for a timer firing.

However, with that fix, during a timer firing I'll get the global window from c.window(), and then I'll be looking up the state in the wrong window if the original element was not in the global window.

I was wondering if maybe I should get the window from the fired timer's namespace? But this seems awkward to do as well because the KeyedWorkItem has multiple timers (timersIterable()) - hmm, should I process them each as if they were separate KeyedWorkItem's?